### PR TITLE
Add server-side Blazor localization tests

### PR DESCRIPTION
**Pull request by rynowak**
This demonstrates a general pattern for Blazor localization, as long with some functional tests (server-side only) that prove that it works.

### CODE REVIEW

File: src/Components/Blazor/Server/src/Builder/BlazorHostingEndpointRouteBuilderExtensions.cs [hunk 1]
```diff
@@ -168,7 +168,19 @@ public static IEndpointConventionBuilder MapFallbackToClientSideBlazor(this IEnd
...
+            // 1. Client app "dist" directory
+            // 2. Client app "wwwroot" directory
+            // 3. Server app "wwwroot" directory
             var directory = endpoints.ServiceProvider.GetRequiredService<IWebHostEnvironment>().WebRootPath;
+            var indexHtml = config.FindIndexHtmlFile();
+            if (indexHtml != null)
+            {
+                directory = Path.GetDirectoryName(indexHtml);
+            }
+
```
Comment (rynowak) [inline src/Components/Blazor/Server/src/Builder/BlazorHostingEndpointRouteBuilderExtensions.cs]: I sent another PR for this - I broke this 😢

File: src/Components/test/testassets/BasicTestApp/Index.razor [hunk 2]
```diff
@@ -60,13 +60,17 @@
...
+        <option value="BasicTestApp.CulturePicker">Culture Picker</option>
+        <option value="BasicTestApp.LocalizedText">Localized Text</option>
     </select>
 
-  @if (SelectedComponentType != null)
-  {
-    <span id="source-info"><code><tt>@(SelectedComponentType.Name.Replace(".", "/")).cshtml</tt></code></span>
-  }
-  <hr />
+    <span id="runtime-info"><code><tt>@System.Runtime.InteropServices.RuntimeInformation.FrameworkDescription</tt></code></span>
```
Comment (rynowak) [inline src/Components/test/testassets/BasicTestApp/Index.razor]: This now shows you if you're running on mono or coreclr. This seemed like a nice improvement to me since I always confuse myself when playing with the test apps.

File: src/Components/test/testassets/BasicTestApp/CulturePicker.razor [hunk 3]
```diff
@@ -0,0 +1,18 @@
+@inject IUriHelper UriHelper
+
+<h3>Select your language</h3>
+<select @onchange="@OnSelected" id="culture-selector">
+    <option>Select...</option>
+    <option value="en-US">English</option>
+    <option value="fr-FR">Français</option>
```
Comment (rynowak) [inline src/Components/test/testassets/BasicTestApp/CulturePicker.razor]: I debated using `en-GB` or `en-CA` or even `en-NZ` - but French seems like a better choice since their number format is significantly different.

The next step here is to add tests that verify number/date parsing.

File: src/Components/test/testassets/TestServer/Controllers/CultureController.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        public IActionResult SetCulture(string culture, string redirectUri)
+        {
+            if (culture != null)
+            {
+                HttpContext.Response.Cookies.Append(
+                    CookieRequestCultureProvider.DefaultCookieName, 
+                    CookieRequestCultureProvider.MakeCookieValue(new RequestCulture(culture)));
+            }
+
+            return LocalRedirect(redirectUri);
```
Comment (rynowak) [inline src/Components/test/testassets/TestServer/Controllers/CultureController.cs]: This is (and the middleware) all that's really required for a server-side app to support localization with the SignalR service. `CookieRequestCultureProvider` is a think we already have.

I didn't include a whole server-side localization scheme in this PR since those are usually app specific.

File: src/Components/test/testassets/TestServer/Startup.cs [hunk 5]
```diff
@@ -63,27 +64,35 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             app.UseAuthentication();
 
             // Mount the server-side Blazor app on /subdir
-            app.Map("/subdir", subdirApp =>
+            app.Map("/subdir", app =>
```
Comment (rynowak) [inline src/Components/test/testassets/TestServer/Startup.cs]: This seemed like a general improvement, but I'm willing to back down if others hate it. It's not valid to reference the *outer* builder here, and this is a way to prevent it from happening (allowed as of C# 8).

File: src/Components/test/testassets/TestServer/Startup.cs [hunk 6]
```diff
@@ -96,6 +105,13 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             {
                 endpoints.MapControllers();
                 endpoints.MapRazorPages();
+
+                // Redirect for convenience when testing locally since we're hosting the app at /subdir/
+                endpoints.Map("/", context =>
+                {
+                    context.Response.Redirect("/subdir");
+                    return Task.CompletedTask;
```
Comment (rynowak) [inline src/Components/test/testassets/TestServer/Startup.cs]: Another *nice improvement* that I wanted to do. I always get upset when I launch the app and get a 404.

**Comment by rynowak** [comment]
Added some basic globalization (number/date parsing) tests. This set can be expanded once we properly support number and date field in non-en-US locales.

File: src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        // see: https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/214
+        //
+        // Calling Clear() can trigger onchange, which will revert the value to its default.
+        private static void ReplaceText(IWebElement element, string text)
+        {
+            element.SendKeys(Keys.Control + "a");
+            element.SendKeys(text);
+        }
```
Comment (rynowak) [inline src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: I'm surprised this problem isn't plaguing us all over with the recent change to how DOM updates works. What's happening here is that when you use `Clear` it will fire `onchange` and set the field to `1/1/0001 ...` then you send the new date value and end up with `1/1/0001 .... 1/2/2000 ...` which isn't valid and gets reset again to `1/1/0001...`.

It makes sense that it behaves this way when you put together all of the pieces, but I've written lots of these tests and its never come up before. Maybe because we're usually working with numbers, and an extra leading `0` is ignored?

Comment (rynowak) [reply src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: And yes, I realize I could have made this nullable, but where's the fun in that?

File: src/Components/test/testassets/BasicTestApp/CulturePicker.razor (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/BasicTestApp/CulturePicker.razor]: C'est un choix judicieux

File: src/Components/test/E2ETest/ServerExecutionTests/LocalizationTest.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.AspNetCore.E2ETesting;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.UI;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.AspNetCore.Components.E2ETest.ServerExecutionTests
+{
+    // For now this is limited to server-side execution because we don't have the ability to set the
+    // culture in client-side Blazor.
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/ServerExecutionTests/LocalizationTest.cs]: I'd have thought that setting the culture for client-side Blazor isn't too troublesome, but we would certainly have issues with loading the satellite assemblies from the `.resx` files.

File: src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj [hunk 9]
```diff
@@ -19,4 +19,17 @@
...
+      <DesignTime>True</DesignTime>
+      <AutoGen>True</AutoGen>
+      <DependentUpon>Resources.resx</DependentUpon>
+    </Compile>
+
+    <EmbeddedResource Update="Resources.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
+  </ItemGroup>
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj]: Just out of interest, is this stuff required? I'm surprised this isn't the default for `resx` within the SDK. Seems a bit inelegant (though not your fault, obviously).

File: src/Components/test/testassets/TestServer/Controllers/CultureController.cs (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/TestServer/Controllers/CultureController.cs]: Looks great. And we already have a server-side mechanism for configuring which set of cultures you support, and what are the fallback rules for unsupported culture requests, right?

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 5 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/TestServer/Startup.cs]: Overall I like it, because many times I've referenced the wrong builder in nested cases and wasted time figuring out why my middleware didn't do the right thing.

File: src/Components/test/testassets/TestServer/Controllers/CultureController.cs (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/TestServer/Controllers/CultureController.cs]: Ah yes I see the config below.

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 6 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/TestServer/Startup.cs]: This doesn't seem quite right to me. If there's a request for `/_framework/something.js` I'd expect it to return a 404, not a redirection to `/subdir`.

Perhaps a better fix would be changing the `launchUrl` in `launchSettings.json`.

File: src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs (hunk 7 above)
Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: I did have to change the E2E bind tests to account for this, because as you point out, `element.Clear()` can no longer truly clear the box if it's bound to a non-nullable number.

I think this is OK because it only affects test cases, not real-world use. In real use, people don't clear a textbox *then* unfocus it *then* refocus it before they start typing. And if they do that, the behavior will still make sense to a human.

Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: > Maybe because we're usually working with numbers, and an extra leading 0 is ignored?

I think the reason is that almost all of our tests only bind things to `string` values. It's only the bind and form-specific tests that go through all the different cases with different types of numbers.

File: src/Components/test/E2ETest/Tests/BindTest.cs [hunk 10]
```diff
@@ -640,5 +641,241 @@ public void CanBindTextboxGenericGuid()
             Browser.Equal(newValue, () => boundValue.Text);
             Assert.Equal(newValue, mirrorValue.GetAttribute("value"));
         }
+
+        // For date comparisons, we parse (non-formatted) values to compare them. Client-side and server-side
+        // Blazor have different formatting behaviour by default.
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Tests/BindTest.cs]: What difference between client-side and server-side are you referring to? Nothing in the test code appears to vary its behavior or expectations based on the runtime type.

**Comment by SteveSandersonMS** [comment]
For the new types of bindable value (`DateTimeOffset`, `DateTime?` + format), is it possible also to have  E2E coverage for invalid (unparseable) data entry cases?

**Comment by SteveSandersonMS** [comment]
This looks really good! At last it feels like we're landing with a properly comprehensive binding system.

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 6 above)
Comment (rynowak) [reply src/Components/test/testassets/TestServer/Startup.cs]: This is using routing, so it's specifically going to match requests to the root URL `/` and only `/`

Comment (rynowak) [reply src/Components/test/testassets/TestServer/Startup.cs]: Map with middleware = map a prefix | Map with routing = match exactly

File: src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs (hunk 7 above)
Comment (rynowak) [reply src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: As a general note, I don't think we had any tests for binding dates before, and I can see why. There's a number of things in this area that are the painful for testing.

File: src/Components/test/E2ETest/ServerExecutionTests/LocalizationTest.cs (hunk 8 above)
Comment (rynowak) [reply src/Components/test/E2ETest/ServerExecutionTests/LocalizationTest.cs]: You could do it in Program.Main (user code) or in the JS code that boots up mono. I spent a little time looking at it but didn't really dig in. If we think it's cheap I'd really like to be able to get it working, but it's also going to require some API design I think.

File: src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj (hunk 9 above)
Comment (rynowak) [reply src/Components/test/testassets/BasicTestApp/BasicTestApp.csproj]: Yes, it doesn't generate the designer file without this.

File: src/Components/test/E2ETest/Tests/BindTest.cs (hunk 10 above)
Comment (rynowak) [reply src/Components/test/E2ETest/Tests/BindTest.cs]: This is because I'm parsing the text into `DateTime` objects and comparing those. The other tests for binding here compare strings directly. 

The differences are things like `01/01/0001` vs `1/1/0001`. ie: just annoying enough to make writing tests hard.

**Comment by rynowak** [comment]
> For the new types of bindable value (DateTimeOffset, DateTime? + format), is it possible also to have E2E coverage for invalid (unparseable) data entry cases?

Sure. I'll probably not have time to get to this until later tonite.

**Comment by rynowak** [comment]
@SteveSandersonMS - updated with some new tests. It would be great if you could take a look at this because I have some other things queued up behind it.

**Review by SteveSandersonMS — APPROVED** [review]
Thanks for figuring out the invalid-value E2E tests!

I expect the formatting differences between server and client should go away as Mono adopts CoreFX fully.
